### PR TITLE
Day4

### PR DESCRIPTION
**Pull request by fstetlerKnightec**

### CODE REVIEW

File: src/main/java/org/advent/dayfour/NumbersPerCard.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+public class NumbersPerCard {
+    private List<Integer> numbersYouHave;
+    private List<Integer> winningNumbers;
+    private int matchingNumbers;
+    private int numberOfTotalCards;
+
+    public NumbersPerCard() {
+    }
```
Comment (Thomas-Zangerl) [inline src/main/java/org/advent/dayfour/NumbersPerCard.java:12]: delete this

File: src/main/java/org/advent/dayfour/DayFour.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    private NumbersPerCard numbersPerCard(String cutString) {
+        NumbersPerCard numbers = new NumbersPerCard();
+
+        String winningString = cutString.split("\\|")[0];
+        String numbersYouHaveString = cutString.split("\\|")[1];
+        List<Integer> listOfWinningNumbers = listOfNumbers(winningString);
+        List<Integer> listOfNumbersYouHave = listOfNumbers(numbersYouHaveString);
+        numbers.setWinningNumbers(listOfWinningNumbers);
+        numbers.setNumbersYouHave(listOfNumbersYouHave);
```
Comment (Thomas-Zangerl) [inline src/main/java/org/advent/dayfour/DayFour.java:69]: this should be in the constructor

File: src/main/java/org/advent/dayfour/DayFour.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    private int totalNumberOfMatchingNumbersPerCard(NumbersPerCard numbersPerCard) {
+        AtomicInteger numbersOfMatchingNumbers = new AtomicInteger();
+        List<Integer> listOfWinning = numbersPerCard.getWinningNumbers();
+        List<Integer> listOfNumbers = numbersPerCard.getNumbersYouHave();
+
+        listOfWinning.forEach(wn -> { if (listOfNumbers.contains(wn)) {numbersOfMatchingNumbers.addAndGet(1);}});
+        return numbersOfMatchingNumbers.get();
+
+    }
```
Comment (Thomas-Zangerl) [inline src/main/java/org/advent/dayfour/DayFour.java:59]: this method takes a single NumbersPerCard object as parameter, which is a strong hint that it rather wants to be a method on that object

File: src/main/java/org/advent/dayfour/DayFour.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        return listOfNumbersPerCard;
+    }
+
+    public List<String> removeFrontPartOfStringAndReturnList(List<String> stringList) {
+        return stringList.stream().map(s -> s.replaceAll(".*:", "")).toList();
+    }
+
+    private int exponentialSumOfNumbersMatchingBetweenOnHandAndWinningPerGame(NumbersPerCard numbersPerCard) {
+        return (int) Math.pow(2, totalNumberOfMatchingNumbersPerCard(numbersPerCard) - 1);
+    }
```
Comment (Thomas-Zangerl) [inline src/main/java/org/advent/dayfour/DayFour.java:49]: should be a method on the object, see comment about totalNumberOfMatchingNumbersPerCard

File: src/main/java/org/advent/dayfour/DayFour.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        AtomicInteger totalSum = new AtomicInteger();
+        listOfNumbersPerCard.forEach(n -> totalSum.addAndGet(n.getNumberOfTotalCards()));
+        return totalSum.get();
+    }
+
+    public int totalSumOfAllExponentialValues(List<NumbersPerCard> listOfNumbersPerCard) {
+        AtomicInteger totalValue = new AtomicInteger();
+        listOfNumbersPerCard.forEach(n -> totalValue.addAndGet(exponentialSumOfNumbersMatchingBetweenOnHandAndWinningPerGame(n)));
+        return totalValue.get();
+    }
```
Comment (Thomas-Zangerl) [inline src/main/java/org/advent/dayfour/DayFour.java:35]: curious to hear the reasoning for this implementation

File: src/main/java/org/advent/dayfour/DayFour.java (hunk 3 above)
Comment (fstetlerKnightec) [reply src/main/java/org/advent/dayfour/DayFour.java:59]: Done in new RP

File: src/main/java/org/advent/dayfour/DayFour.java (hunk 5 above)
Comment (fstetlerKnightec) [reply src/main/java/org/advent/dayfour/DayFour.java:35]: Fixed to         return listOfNumbersPerCard.stream().mapToInt(npc -> npc.exponentialSumOfNumbersMatchingBetweenOnHandAndWinningPerGame(npc)).sum();

File: src/main/java/org/advent/dayfour/DayFour.java (hunk 4 above)
Comment (fstetlerKnightec) [reply src/main/java/org/advent/dayfour/DayFour.java:49]: done

File: src/main/java/org/advent/dayfour/DayFour.java (hunk 2 above)
Comment (fstetlerKnightec) [reply src/main/java/org/advent/dayfour/DayFour.java:69]: Done in new PR

File: src/main/java/org/advent/dayfour/NumbersPerCard.java (hunk 1 above)
Comment (fstetlerKnightec) [reply src/main/java/org/advent/dayfour/NumbersPerCard.java:12]: Didnt delete, put up new parameters in and added setters

**Pull request closed without merging**